### PR TITLE
DataViews Sidebar: Display item count on DataViews sidebar

### LINKED ISSUE #57072: Data views: Show total count in menu item
In the Patterns page, each view (pattern category) has an associated count:

Screenshot 2023-12-14 at 16 22 10

It would be nice to bring this UI detail to other views, e.g.:

Screenshot 2023-12-14 at 16 21 32

### PR DESCRIPTION
**Pull request by ramonjd**
Resolves https://github.com/WordPress/gutenberg/issues/57072

## What?

Displays item count on DataViews sidebar.

> [!NOTE]
> Props to @Souptik2001 for starting this work, and also to the following contributors: https://github.com/WordPress/gutenberg/pull/62028#issuecomment-2133482231 Please include these handles when merging.
 

## Why?

From https://github.com/WordPress/gutenberg/pull/62028

> In the Pattern management sidebar each category is represented by a menu item. Inside each menu item is a count indicating how many patterns there are in the category.

> Therefore, it would be good to also display that on the "Pages" management sidebar, so that anyone can get the count of each view without visiting it.

## How?

From https://github.com/WordPress/gutenberg/pull/62028

> Display the page count beside each view.
> Achieved by fetching all the pages and then doing a filter using Javascript.

## Testing Instructions

1. Go to site-editor.
2. Go to "Pages".
3. On the sidebar, you will see that you have the count of each item/view.

## Screenshots or screencast 

Screenshot 2024-09-11 at 12 47 16 PM

### CODE REVIEW

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/index.js]: > Is the span needed?

Asked here: https://github.com/WordPress/gutenberg/pull/62028#discussion_r1618917031

Yes, otherwise the content will not render as a child of `<HStack />`

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/default-views.js]: I thought a follow up could filter out items that have no counts, e.g., `.filter( ( { count } ) => count > 0 )`


https://github.com/user-attachments/assets/796a0971-0949-43c0-a247-40a9bc0a68cb

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/default-views.js]: I pulled this filter across from the original PR, but it effectively does nothing when you disable the filter in the view.

I think we should remove it.

The [query arg builder](https://github.com/WordPress/gutenberg/blob/17c1206bc169b9a186507fef6608d53e4aa7f602/packages/edit-site/src/components/post-list/index.js#L207-L207) in `<PostList />` doesn't seem to take excluded statuses into account.

As mentioned by @Souptik2001  in https://github.com/WordPress/gutenberg/pull/62028, I think the [default statuses should include "trash"](https://github.com/WordPress/gutenberg/blob/79ff3873ed8c1a2d1af9b7f9bc25416023adc801/packages/edit-site/src/components/post-list/index.js#L174-L174) because the "All pages" count includes the trashed items count but doesn't display the items themselves. 

Either that, or we don't add trashed items to the count and do a separate query for "trashed" items.

Regardless, there seems to be a bug when activating/deactivating fields. No records are returned and therefore [getItemId](https://github.com/WordPress/gutenberg/blob/6a60081909d1dfbbb608e7b069dec3eb6881e37e/packages/edit-site/src/components/post-list/index.js#L176-L176) triggers an error

https://github.com/WordPress/gutenberg/blob/6a60081909d1dfbbb608e7b069dec3eb6881e37e/packages/edit-site/src/components/post-list/index.js#L244-L249

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/default-views.js:74]: I probably followed the [same rabbit hole](https://github.com/WordPress/gutenberg/pull/62028#discussion_r1616054976) as @Souptik2001 here. 

Unfortunately I couldn't find another way either. 

`getEntityRecordsTotalItems` won't return anything unless there's been a `getEntityRecords` query to populate the state, so we have to fetch the items.

Ideally, I guess, there'd be an options request with a limited database query `SELECT COUNT` to set the `X-WP-Total` header or something... I'm not sure

**Review by ramonjd — COMMENTED** [review]
Rebased version of https://github.com/WordPress/gutenberg/pull/62028 based on the latest in trunk.

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: > No records are returned and therefore [getItemId](https://github.com/WordPress/gutenberg/blob/6a60081909d1dfbbb608e7b069dec3eb6881e37e/packages/edit-site/src/components/post-list/index.js#L176-L176) triggers an error

PR to fix that I think:

https://github.com/WordPress/gutenberg/pull/65227

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (jameskoster) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: Hiding empty bundled views seems reasonable. Custom views should probably remain visible regardless.

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (jameskoster) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: > disable the filter in the view.

Ideally this should not be possible. See https://github.com/WordPress/gutenberg/issues/60468.

If possible I don't think the "All pages" view should include trash; that's not how it works in wp-admin.

**Comment by jameskoster** [comment]
Thanks for picking this up :) 

Could we update the spacing to match the Patterns page?

| Patterns | Pages |
| --- | --- |
| Screenshot 2024-09-11 at 11 57 01 | Screenshot 2024-09-11 at 11 56 51 |

I assume adding counts to Custom views would be a separate effort? How about views in the Templates section?

**Review by DaniGuardiola — APPROVED** [review]
Thanks for doing the rebasing work. LGTM, though let's wait for any pending discussions to resolve before merging as I'm not super familiar with the details.

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: Custom views are unaffected. There is a snag however: it takes a tick longer to fetch the counts so you'll see a little delay on first load.

https://github.com/user-attachments/assets/6555f459-fcf7-4df1-aa00-ab4a06a7dcf6

A few options I can think of, some better than others:

1. Don't filter out `0` values. That means the list will appear in full, but the counts, including `0` items, will appear a little later. Maybe the `0` item could be unclickable/disabled/greyed out.
2. Show the full list, then filter out `0` values. That means the full list will appear, but `0` value items will _disappear_. 👎🏻 
3. Include a loading placeholder.
4. ?

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: Okay, I left it as it is (All pages do not include"trashed" items) but removed the filter.

**Comment by ramonjd** [comment]
Thanks for testing, folks!

> I assume adding counts to Custom views would be a separate effort? How about views in the Templates section?

I'm not yet familiar enough with the views to give a 100% certain answer, but my 99% hunch is that yes, it'd require another piece of work.

> Could we update the spacing to match the Patterns page?

👍🏻 

To explain the discrepancy:

`suffix` according to `<DataViewItem />` is content _outside_ `<SidebarNavigationItem />`.

`<SidebarNavigationItem />` however takes it own `suffix` prop, which appears in the button item. That's where the correct padding comes from.

https://github.com/WordPress/gutenberg/blob/eae814316bb04260272063e18226f19f6b89b540/packages/edit-site/src/components/sidebar-dataviews/dataview-item.js#L56-L64

File: packages/edit-site/src/components/sidebar-dataviews/dataview-item.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/dataview-item.js:59]: I added a new prop because:

- [SidebarNavigationItem](https://github.com/WordPress/gutenberg/blob/eae814316bb04260272063e18226f19f6b89b540/packages/edit-site/src/components/sidebar-navigation-item/index.js#L79-L79) takes a suffix prop that we need. We want the incoming content to be clickable.
- The ["outside" suffix](https://github.com/WordPress/gutenberg/blob/eae814316bb04260272063e18226f19f6b89b540/packages/edit-site/src/components/sidebar-dataviews/dataview-item.js#L64-L64) prop is used for the [drop down menu](https://github.com/WordPress/gutenberg/blob/eae814316bb04260272063e18226f19f6b89b540/packages/edit-site/src/components/sidebar-dataviews/custom-dataviews-list.js#L115-L115). So it should really sit outside the clickable button.

What might be a better refactor after this PR, is for `<DataViewItem />` to take a sidebar item props object or something like that, which only applies to the child `<SidebarNavigationItem />`  e.g., 

```jsx

<DataViewItem
    navigationItemProps={
        suffix: ...,
        isActive: ...,
        ...
     } />

```

File: packages/edit-site/src/components/sidebar-dataviews/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/index.js:22]: I created a new hook because `useDefaultViews()` is used in several places and not all of them require fetching the count.

File: packages/edit-site/src/components/sidebar-dataviews/style.scss
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/style.scss:20]: This is to make sure the rhs padding matches the pattern view's items.

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/default-views.js:85]: The list will load faster than the count, so let's return that while we wait.

**Review by ramonjd — COMMENTED** [review]
Grabbing the counts is a tad resource intensive. 

I've been testing with a test with 100+ pages. It's not too bad, but who knows how it will fare with sites with 1000s of pages.

I haven't looked too deeply, but I'd guess there's a bit of effort involved creating a more performant count. We could probably create a custom endpoint to return `$wpdb->get_var("SELECT COUNT(*)...`

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: I've done 1 for now:

https://github.com/user-attachments/assets/4058ecb2-4d9c-47e0-9a48-de4591aab457

Comment (jameskoster) [reply packages/edit-site/src/components/sidebar-dataviews/default-views.js]: A loading indicator would be a nice enhancement but can potentially come later.

**Comment by jameskoster** [comment]
In terms of visuals I reckon this is in a decent spot. The only outstanding issue I see is the inclusion of `trash`ed pages in the All Pages view which produces a mismatch in counts. Compare the menu item with the running total in the footer:

Screenshot 2024-09-12 at 15 27 42

Ideally we exclude trashed items from the All pages count.

**Comment by ramonjd** [comment]
> Ideally we exclude trashed items from the All pages count.

Oh yeah, we could minus the number of trashed pages from the total count. I'll add it in. 🍺

File: packages/edit-site/src/components/sidebar-dataviews/default-views.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-dataviews/default-views.js:105]: Remove trash items from the all items count as per the default view (e.g., for pages: https://github.com/WordPress/gutenberg/blob/71517b48929e0678066efe1ca8dcd3cc793dfb8d/packages/edit-site/src/components/post-list/index.js#L174-L174)

**Comment by ramonjd** [comment]
Are folks okay to have this merged for now?

**Comment by ramonjd** [comment]
I'll merge and keep an eye on any post-merge feedback 👍🏻

**Comment by youknowriad** [comment]
Hey folks! I understand that a level of efforts went into this PR but I would rather not have this feature than do unlimited page requests like this (for pages, posts...). It's fine for templates.

As an side, this PR increased the "load pages" perf metric numbers.

**Comment by ramonjd** [comment]
> Hey folks! I understand that a level of efforts went into this PR but I would rather not have this feature than do unlimited page requests like this (for pages, posts...). It's fine for templates.

@youknowriad Thanks for the heads up. I'll revert today. 

I agree. Until we come up with an alternative, e.g., `SELECT COUNT` response [mentioned above](https://github.com/WordPress/gutenberg/pull/65223#discussion_r1753067847) probably best to leave this. 

Noting that a get posts/entity count might be useful elsewhere in the future I suppose, e.g., "You have a billion attachments in your media library" 😄

**Comment by ramonjd** [comment]
Revert PR: https://github.com/WordPress/gutenberg/pull/65625

**Comment by mtias** [comment]
@ramonjd curious if we should have an endpoint that just does counts for post types or expose total counts when fetching an entity regardless of the paged / per page setting.

**Comment by ramonjd** [comment]
@mtias Good question. 

I let the cat out of the bag in relation to the custom endpoint [above](https://github.com/WordPress/gutenberg/pull/65223#pullrequestreview-2299163013), but it would be also useful I imagine to have status counts exposed in entity requests as well.

While `X-WP-Total` is available in the headers for single entity requests, for this use case, and given the current set up, there'd have be a separate request for each status to grab the totals. There could be `n` [statuses](https://developer.wordpress.org/reference/functions/register_post_status/). 

What would be ideal, though I haven't looked too deeply into it, is one response for any given post type with counts for each status. It could be a route fragment with an optional field, e.g.,  `posts/counts?_fields=status`. 

Furthermore, if we make it embeddable in the regular `/posts` response, consumers could have one or the other — entity collection or just counts — or both.

**Comment by ramonjd** [comment]
> curious if we should have an endpoint that just does counts for post types or expose total counts when fetching an entity regardless of the paged / per page setting.

Looks like I forgot about [wp_count_posts](https://developer.wordpress.org/reference/functions/wp_count_posts/) 😄 

I have a draft PR to create an endpoint around this: 

- https://github.com/WordPress/gutenberg/pull/66294

**Comment by jeffpaul** [comment]
@finitelydope mind sharing your WordPress.org username so I can ensure you appear in the 6.7 credits listing?

**Comment by ramonjd** [comment]
This PR was reverted in https://github.com/WordPress/gutenberg/pull/65625, but I think it can be revived if we have a more performant way to fetch the counts. PR for that here: https://github.com/WordPress/gutenberg/pull/66294

**Pull request merged**
